File: sdk/ios/Sources/FrakSDK/Tracking/PurchaseSender.swift
import Foundation

/// `POST /user/track/purchase`. `QueuedRow.payload` is the wire body minus `merchantId`, filled in here.
struct PurchaseSender: RowSender {
    static let kind = "purchase"
    private static let path = "/user/track/purchase"

    func deliver(row: QueuedRow, ctx: SendContext) async throws(CancellationError) -> DeliveryOutcome {
        let merchantId: String
        if let known = row.merchantId {
            merchantId = known
        } else if let resolved = await ctx.resolveMerchantId() {
            merchantId = resolved
        } else {
            return .hold
        }
        guard let built = RowBody.withMerchantId(row.payload, merchantId: merchantId) else {
            return .dropped
        }

        do {
            let response = try await ctx.http.post(Self.path, body: built.data, headers: clientIdHeaders(row))
            return classify(response)
        } catch let error as FrakError {
            return .retryable(error)
        } catch {
            // Cancellation, and nothing else — HTTPClient maps every transport failure to a FrakError.
            throw CancellationError()
        }
    }
}

File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/tracking/PurchaseSender.kt
package id.frak.sdk.tracking

import id.frak.sdk.core.FrakError
import org.json.JSONObject

/** `POST /user/track/purchase`. [QueuedRow.payload] is the wire body minus `merchantId`, filled in here. */
internal class PurchaseSender : RowSender {
    override suspend fun deliver(
        row: QueuedRow,
        ctx: SendContext,
    ): DeliveryOutcome {
        val merchantId = row.merchantId ?: ctx.resolveMerchantId() ?: return DeliveryOutcome.Hold
        val body = JSONObject(row.payload.toString()).put(MERCHANT_ID_KEY, merchantId)

        val response =
            try {
                ctx.http.post(PATH, body.toString(), clientIdHeaders(row))
            } catch (failure: FrakError) {
                return DeliveryOutcome.Retryable(failure)
            }

        return classifyStatus(response)
    }

    companion object {
        const val KIND = "purchase"
        const val PATH = "/user/track/purchase"

        private const val MERCHANT_ID_KEY = "merchantId"
    }
}
